frontend/gateway/grpcclient/client.go

```diff
@@ -484,7 +484,31 @@ func (c *grpcClient) Solve(ctx context.Context, creq client.SolveRequest) (res *
 
 func (c *grpcClient) ResolveSourceMetadata(ctx context.Context, op *opspb.SourceOp, opt sourceresolver.Opt) (*sourceresolver.MetaResponse, error) {
 	if c.caps.Supports(pb.CapSourceMetaResolver) != nil {
-		return nil, errors.Errorf("fallback not implemented")
+		var ref string
+		if strings.HasPrefix(op.Identifier, "docker-image://") {
+			ref = strings.TrimPrefix(op.Identifier, "docker-image://")
+		} else if strings.HasPrefix(op.Identifier, "oci-layout://") {
+			ref = strings.TrimPrefix(op.Identifier, "oci-layout://")
+		} else {
+			return &sourceresolver.MetaResponse{Op: op}, nil
+		}
+		retRef, dgst, config, err := c.ResolveImageConfig(ctx, ref, opt)
+		if err != nil {
+			return nil, err
+		}
+		if strings.HasPrefix(retRef, "docker-image://") {
+			op.Identifier = "docker-image://" + retRef
+		} else if strings.HasPrefix(retRef, "oci-layout://") {
+			op.Identifier = "oci-layout://" + retRef
```

Comment:
If `HasPrefix`, prepend it again? I assume this should be `op.Identifier`.

Reply:
I think this is still wrong. `ref` has had the prefix stripped already, so the condition never triggers.

This should be:

```go
if strings.HasPrefix(op.Identifier, "docker-image://") {
	op.Identifier = "docker-image://" + retRef
}
```